connectors/utils.py

```diff
@@ -396,6 +396,20 @@ async def put(self, coroutine, result_callback=None):
         )
         return task
 
+    async def try_put(self, coroutine, result_callback=None):
+        """Tries to add a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function return a None task immediately
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+
+        if self._sem.locked():
+            return None
+
+        return await self.put(coroutine, result_callback=result_callback)
```

Comment:
IMO there is a potential for a race condition here, if the semaphore becomes locked between the check `if self._sem.locked():` and the `return await self.put(..)`. 

I think it should still be safe because it essentially means that we will be less efficient at handling tasks in job execution service. Is there any potential for unwanted impact in this race condition scenario?

Reply:
+++ to addressing this. I think it's possible that once `await self.put` is called, execution is given to another coroutine that can in turn call `try_put`.

Since semaphore is set inside `self.put`, it won't yet be set in this case for both calls and both will go through.

Reply:
First of all, I want to make it clear that `ConcurrentTasks` is not designed to be thread-safe. [asyncio.BoundedSemaphore](https://docs.python.org/3/library/asyncio-sync.html#boundedsemaphore) is not either.

But I think your concern is valid. I assumed it would pass the control back to the main loop when it encountered the first `await` in `put` method (i.e. `await self._sem.acquire()`), but I coudn't find the evidence.

I will make some changes to handle this case.

Reply:
I tested with the following script:
```python
import asyncio
from datetime import datetime

async def put(sem, i):
  print(f"{datetime.now()} - putting {i}")
  # use asyncio.sleep(0) to release control
  await asyncio.sleep(0)
  await sem.acquire()
  print(f"{datetime.now()} - Successfully put {i}")

  # sleep for 2 second and release
  await asyncio.sleep(2)
  print(f"{datetime.now()} - Successfully done {i}")
  sem.release()

async def try_put(sem, i):
  # wait for other tasks
  await asyncio.sleep(0.5)
  print(f"{datetime.now()} - try_put {i}")
  if sem.locked():
    print(f"{datetime.now()} - try_put {i}, couldn't put the task, return")
    return

  await put(sem, i)


async def main():
  task_pool_limit = 5
  num_of_tasks = task_pool_limit + 1
  sem = asyncio.BoundedSemaphore(task_pool_limit)
  tasks = []
  for i in range(num_of_tasks):
    tasks.append(asyncio.create_task(try_put(sem, i+1)))

  await asyncio.gather(*tasks)

asyncio.run(main())

```
and here is the output:
```
2023-11-09 17:54:40.309128 - try_put 1
2023-11-09 17:54:40.309180 - putting 1
2023-11-09 17:54:40.309209 - try_put 2
2023-11-09 17:54:40.309221 - putting 2
2023-11-09 17:54:40.309237 - try_put 3
2023-11-09 17:54:40.309247 - putting 3
2023-11-09 17:54:40.309266 - try_put 4
2023-11-09 17:54:40.309276 - putting 4
2023-11-09 17:54:40.309291 - try_put 5
2023-11-09 17:54:40.309300 - putting 5
2023-11-09 17:54:40.309314 - try_put 6
2023-11-09 17:54:40.309323 - putting 6 --> trying to put task 6, blocking instead of returning
2023-11-09 17:54:40.309365 - Successfully put 1
2023-11-09 17:54:40.309396 - Successfully put 2
2023-11-09 17:54:40.309419 - Successfully put 3
2023-11-09 17:54:40.309440 - Successfully put 4
2023-11-09 17:54:40.309460 - Successfully put 5
2023-11-09 17:54:42.310391 - Successfully done 1
2023-11-09 17:54:42.310486 - Successfully done 2
2023-11-09 17:54:42.310520 - Successfully done 3
2023-11-09 17:54:42.310548 - Successfully done 4
2023-11-09 17:54:42.310574 - Successfully done 5
2023-11-09 17:54:42.310632 - Successfully put 6 --> task 6 is successfully put after blocking for 2 seconds
2023-11-09 17:54:44.311775 - Successfully done 6
```

But if I comment out `await asyncio.sleep(0)`, it works as expected:
```
2023-11-09 17:56:52.230397 - try_put 1
2023-11-09 17:56:52.230523 - putting 1
2023-11-09 17:56:52.230562 - Successfully put 1
2023-11-09 17:56:52.230647 - try_put 2
2023-11-09 17:56:52.230688 - putting 2
2023-11-09 17:56:52.230710 - Successfully put 2
2023-11-09 17:56:52.230768 - try_put 3
2023-11-09 17:56:52.230809 - putting 3
2023-11-09 17:56:52.230823 - Successfully put 3
2023-11-09 17:56:52.230857 - try_put 4
2023-11-09 17:56:52.230873 - putting 4
2023-11-09 17:56:52.230886 - Successfully put 4
2023-11-09 17:56:52.230917 - try_put 5
2023-11-09 17:56:52.230932 - putting 5
2023-11-09 17:56:52.230945 - Successfully put 5
2023-11-09 17:56:52.230973 - try_put 6
2023-11-09 17:56:52.230987 - try_put 6, couldn't put the task, return --> trying to put task 6, return immediately
2023-11-09 17:56:54.231842 - Successfully done 1
2023-11-09 17:56:54.231925 - Successfully done 2
2023-11-09 17:56:54.231960 - Successfully done 3
2023-11-09 17:56:54.231992 - Successfully done 4
2023-11-09 17:56:54.232019 - Successfully done 5
```

So it seems the current implementation does not release control to the loop unless you do so explicitly. 

I don't find an easy way to prevent this from happening. Considered using [asyncio.Lock](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock), so that there can only be one party executing `put`/`try_put`, so that the context switching won't happen between the lock is acquired and released. But this will not work when a user calls `put` and is blocked, then another user calls `try_put` and will be blocked again because it can't acquire the lock.

Reply:
I think I found an easy workaround. WDYT if we extend the `asyncio.BoundedSemaphore` class and introduce a `try_acquire` method:

```python
def try_acquire(self):
  if self.locked():
    return False

  self._value -= 1
  return True
```

The `locked` implementation is [here](https://github.com/python/cpython/blob/97c4c06d0d235aad00e5b6b10af8b8d68c889b9b/Lib/asyncio/locks.py#L358-L361). It will return `True` if either there's no space or there are other tasks blocking to acquire it.

The code
```
self._value -= 1
return True
```
is borrowed from the implementation of [acquire](https://github.com/python/cpython/blob/97c4c06d0d235aad00e5b6b10af8b8d68c889b9b/Lib/asyncio/locks.py#L363). The logic for `try_acquire` is simple because we don't need to add it to the waiting pool when there's no space available immediately.

Reply:
Intuitively it looks okay, but I need to think a bit more about it, thanks for sharing the thought process!

Reply:
I implemented the above change in a separate PR: #1889

Reply:
I've tried an approach with keeping the `Event` working and it seemed to be a little cleaner:

```python
+    async def try_put(self, coroutine, result_callback=None):
+        """Adds a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function will block and wait for a free slot.
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+        # If self.tasks has reached its max size, we wait for one task to finish
+        if len(self.tasks) >= self.max_concurrency:
+            try:
+                await asyncio.wait_for(self._task_over.wait(), 1) # Wait for 1 second and if we can't add a task error out
+                self._task_over.clear()
+            except asyncio.TimeoutError:
+                return None
+        # rearm
+        task = asyncio.create_task(coroutine())
+        self.tasks.append(task)
+        task.add_done_callback(
+            functools.partial(self._callback, result_callback=result_callback)
+        )
+        return task
```

Reply:
Do you think it can work?

Reply:
Replied in anther PR: https://github.com/elastic/connectors/pull/1889#discussion_r1395312172, to make discussions in a single place.